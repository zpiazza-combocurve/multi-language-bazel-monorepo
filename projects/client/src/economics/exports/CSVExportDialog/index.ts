import CustomReportDialogTheme from './theme';

export * from './helpers';
export * from './types';
export * from './Tour';

export { CustomReportDialogTheme };
