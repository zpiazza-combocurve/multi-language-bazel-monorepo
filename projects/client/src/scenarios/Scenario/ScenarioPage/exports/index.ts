export { ExportToAriesDialog } from './ExportToAriesDialog';
export { ExportToCsvDialog } from './ExportToCsvDialog';
export { ExportToPhdwinDialog } from './ExportToPhdwinDialog';
export * from './ExportProbabilisticDialog';
export { default as CCImportDialog } from './CCImportDialog';
export * from './CCExportDialog';
export * from './useCCExport';
