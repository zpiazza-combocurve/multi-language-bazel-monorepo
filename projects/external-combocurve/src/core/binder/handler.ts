/* eslint-disable complexity, @typescript-eslint/ban-types */
import 'reflect-metadata';
import { Request } from 'express';

import { RequestStructureError, ValidationError } from '@src/helpers/validation';

import { FieldMetadata, MetadaWrapper } from '../metadata/metada-wrapper';
import { IInternalRequestBindOptions, NamingTypes, ParseOutput } from '../common';

import { appParses } from './parses';
import { RequestStructure } from './request-structure';
import { sourceFieldLookup } from './field-lookup';

const propertyNotFound = Symbol('propertyNotFound');

// Just a helper type to avoid write Record<string, unknown> everywhere
type JsObj = Record<string, unknown>;

export class BindHandler<TRequestType extends {}> {
	errors: ValidationError[] = [];

	constructor(
		private requestModel: TRequestType,
		private request: Request,
		private metadata: MetadaWrapper,
		private requestStructure: RequestStructure,
		private namingConfig = NamingTypes.ExactlyEqual,
	) {}

	public getModel(): TRequestType {
		return this.requestModel;
	}

	public bind(): ValidationError[] {
		this.innerBind('', this.requestModel, this.metadata.fields);
		return this.errors;
	}

	private innerBind(
		location: string,
		modelAux: JsObj,
		fields: FieldMetadata[],
		scopedSource: unknown = undefined,
	): void {
		// Check unknown fields on source
		if (scopedSource !== undefined) {
			this.errors.push(
				...this.requestStructure.checkUnknowInfoOnSource(
					scopedSource as Record<string, unknown>,
					fields.map((f) => f.name),
					location,
				),
			);
		}

		for (const field of fields) {
			if (field.options.fromWhere === 'composition' && field.innerFields) {
				if (modelAux[field.name] === undefined) {
					modelAux[field.name] = getInstance(field.options.objFactory);
				}

				const innerModel = modelAux[field.name] as JsObj;
				this.innerBind(location, innerModel, field.innerFields.fields, undefined);

				continue;
			}

			const requestValue = this.getPropertyFromRequest(field, scopedSource);

			// If the property is not found, it's not a problem if the property is optional
			if (requestValue === propertyNotFound) {
				if (field.options.isOptional !== true) {
					this.errors.push(new RequestStructureError(`Property ${field.name} not found`, location));
				}

				continue;
			}

			// When object is expected, It's create a scoped that is the object value from request
			// And then called this same bind recursively to fill the object using this new scope
			if (field.options.expects === 'object' && field.options.customParse === undefined && field.innerFields) {
				if (typeof requestValue !== 'object') {
					this.errors.push(new RequestStructureError(`Property ${field.name} is not an object`, location));
				}

				modelAux[field.name] = getInstance(field.options.objFactory);

				this.innerBind(
					location + field.name,
					modelAux[field.name] as JsObj,
					field.innerFields.fields,
					requestValue,
				);

				continue;
			}

			let parsedValue = requestValue;

			// We don't need to parse services since they are come from our context
			if (field.options.expects !== 'service') {
				parsedValue = this.getPropertParsedValue(field, requestValue, location);
			}

			modelAux[field.name] = parsedValue;
		}
	}

	private getPropertyFromRequest(field: FieldMetadata, scopedSource: unknown = undefined): unknown {
		const source = this.requestStructure.getRequestSource(field.options.fromWhere, scopedSource);

		// We accept both array and single object for body
		if (field.options.fromWhere === 'body' && field.options.isBody) {
			return field.options.expects === 'array' && !Array.isArray(source) ? [source] : source;
		}

		const keyValue = sourceFieldLookup(source, field.name, this.namingConfig);
		return keyValue?.value ?? propertyNotFound;
	}

	private getPropertParsedValue(
		field: FieldMetadata,
		requestValue: unknown,
		location: string,
		overrideExpects?: string,
	): unknown {
		const expectType = overrideExpects ?? field.options.expects;

		if (expectType === 'array') {
			return this.getArrayPropertParsedValue(field, requestValue, location);
		}

		let parseFN = appParses[expectType];
		if (field.options.customParse !== undefined) {
			parseFN = field.options.customParse;
		}

		const output = parseFN({
			name: field.name,
			value: requestValue,
			location,
			model: this.requestModel,
			request: this.request,
			requirements: field.options.requirements,
		});
		if (output.errorMsg) {
			this.errors.push(
				output.errorName
					? new ValidationError(output.errorMsg, location, output.errorName)
					: new RequestStructureError(output.errorMsg, location),
			);
			return;
		}

		this.checkTargetConstructor(field.options, output, location);
		return output.parsedValue;
	}

	// Fill Array
	private getArrayPropertParsedValue(field: FieldMetadata, requestValue: unknown, location: string): unknown {
		if (!Array.isArray(requestValue)) {
			if (field.options.fromWhere === 'query') {
				requestValue = [requestValue];
			} else {
				this.errors.push(new RequestStructureError(`Property ${field.name} is not an array`, location));
				return;
			}
		}

		const aux = requestValue as unknown[];
		const minLength = field.options.requirements?.minItems ?? 0;
		const maxLength = field.options.requirements?.maxItems ?? Number.MAX_SAFE_INTEGER;

		if (aux.length < minLength || aux.length > maxLength) {
			this.errors.push(
				new RequestStructureError(
					`${field.name} must have between ${minLength} and ${maxLength} items`,
					location,
				),
			);
			return;
		}

		if (!field.options.itemsExpects) {
			this.errors.push(
				new RequestStructureError(`Items expects not defined for property array ${field.name}`, location),
			);
			return;
		}

		if (field.options.itemsExpects === 'object' && !field.options.objFactory) {
			this.errors.push(
				new RequestStructureError(`typeFactory defined for property object ${field.name}`, location),
			);
			return;
		}

		let arrayIdx = 0;
		const output: unknown[] = [];

		for (const item of aux) {
			const currentLocation = `${location}${field.options.isBody ? '' : field.name}[${arrayIdx++}]`;
			if (field.options.itemsExpects === 'object' && field.innerFields) {
				const obj = getInstance(field.options.objFactory);
				output.push(obj);

				this.innerBind(currentLocation, obj as JsObj, field.innerFields.fields, item);
			} else {
				output.push(this.getPropertParsedValue(field, item, currentLocation, field.options.itemsExpects));
			}
		}

		return output;
	}

	private checkTargetConstructor(options: IInternalRequestBindOptions, parsedValue: ParseOutput, location: string) {
		if (options.targetConstructor === undefined) {
			return;
		}

		try {
			parsedValue.parsedValue = options.targetConstructor(parsedValue.parsedValue);
		} catch (err) {
			const error = err as Error;
			this.errors.push(new ValidationError(error.message, location, error.name));
		}
	}

	public hasErrors(): boolean {
		return true;
	}
}

function getInstance(factory: (() => unknown) | undefined): unknown {
	let aux: unknown = factory;
	while (typeof aux === 'function') {
		aux = aux();
	}

	return aux;
}
