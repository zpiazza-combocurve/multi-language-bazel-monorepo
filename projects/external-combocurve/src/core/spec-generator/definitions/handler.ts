/* eslint-disable 
	complexity,
	@typescript-eslint/explicit-module-boundary-types, 
	@typescript-eslint/no-explicit-any,
	@typescript-eslint/no-non-null-assertion */
import Handlebars from 'handlebars';

import { definitionsTemplate, ObjectDefinitionSpec } from '../templates';
import { IInternalRequestBindOptions, RequestBindOptionsKey, RequestBindPropertiesKey } from '../../common';
import { Controller } from '../../controllers/base';
import { fillDefinitionExamples } from '../example.helper';
import { PayloadCommandRequest } from '../../requests/base';
import { specs } from '../../metadata/metadata';

import { reqParseInput, requestSpecParses, resParseInput, responseSpecParses } from './parses';

type sampleObjContext = {
	obj: any; // The object sample
	name?: string; // The name of the definitions. It's different for reponse and inner objects
	parentKey?: string; // When it's a inner obj of a response, this is the key of the parent object
};

const statusNameMap: Record<number, string> = {
	200: 'OK',
	201: 'Created',
	202: 'Accepted',
	204: 'NoContent',
	207: 'MultiStatus',
	400: 'BadRequest',
	401: 'Unauthorized',
	403: 'Forbidden',
	404: 'NotFound',
};

export class DefinitionsGenHandler<TController extends Controller> {
	private controller: TController;
	private definitionsBasePath: string;

	constructor(controller: TController, definitionsBasePath?: string) {
		this.controller = controller;
		this.definitionsBasePath = definitionsBasePath ?? '#/definitions/';
	}

	public getDefinitionSpec(): string {
		const requestHash = new Set<string>();
		const objectsDefinitions: ObjectDefinitionSpec[] = [];

		for (const route of this.controller.Routes) {
			const routeRequestModel = route.requestSample;
			const requestName = this.getNameFromType(routeRequestModel);

			// Sometimes we can use the same request model for different routes
			if (requestHash.has(requestName)) {
				continue;
			}

			requestHash.add(requestName);

			objectsDefinitions.push(...this.mapRequestObjects(routeRequestModel));
			objectsDefinitions.push(...this.mapResponseSamples(requestName, routeRequestModel as Record<string, any>));
		}

		// Create the complete examples
		fillDefinitionExamples(objectsDefinitions);

		this.addDefinitionsBasePath(objectsDefinitions);

		const template = Handlebars.compile(definitionsTemplate);
		const result = template({
			objects: objectsDefinitions.filter((f) => f.properties.length > 0),
		});

		// The template has a empty line at the beginning, so we need to remove it
		return result.substring(result.indexOf('\n') + 1);
	}

	private addDefinitionsBasePath(objectsDefinitions: ObjectDefinitionSpec[]): void {
		for (const obj of objectsDefinitions) {
			for (const objProp of obj.properties.filter((p) => p.typeKey === '$ref')) {
				objProp.type = this.definitionsBasePath + objProp.type;
			}

			for (const objProp of obj.properties.filter((p) => p.itemsTypeKey === '$ref')) {
				objProp.type = this.definitionsBasePath + objProp.itemsType;
			}
		}
	}

	public mapRequestObjects(request: any): ObjectDefinitionSpec[] {
		// Run through request's properties and get all objects
		// This is pretty much a single graph traversal

		const output: ObjectDefinitionSpec[] = [];
		const objects: any[] = [request];

		while (objects.length > 0) {
			const currentObject = objects.pop();
			const objectSpec: ObjectDefinitionSpec = {
				name: this.getNameFromType(currentObject),
				properties: [],
				required: [],
			};

			const bindProperties = Reflect.getMetadata(RequestBindPropertiesKey, currentObject) as string[];

			for (const propertyKey of bindProperties ?? []) {
				const options = Reflect.getMetadata(
					RequestBindOptionsKey,
					currentObject,
					propertyKey,
				) as IInternalRequestBindOptions;

				// Spec definitions just consider body objects
				if (!options || (options.fromWhere !== 'body' && options.fromWhere !== 'scope')) {
					continue;
				}

				// CommandRequest has a special treatment, because it's have a payload property that should be ignored
				if (options.isBody && this.isCommandRequest(currentObject, options) && options.objFactory) {
					objects.push(options.objFactory());
					continue;
				}

				if (options.expects === 'object') {
					// We have a unique edge case that's when the propert is a object but have a sample.
					// In general this means the object does not have a bunch of these decorators
					// so we can't get the properties from metadata
					if (options.specOptions?.sample) {
						this.requestSampleEdgeCase(propertyKey, options, objectSpec, output);
						continue;
					}

					objects.push(options.objFactory!());
				}

				const specParse = requestSpecParses[options.expects];
				if (specParse) {
					const input: reqParseInput = {
						prop: propertyKey,
						opt: options,
					};

					objectSpec.properties.push(specParse(input));
				}

				if (options.isOptional !== true) {
					objectSpec.required!.push(propertyKey);
				}
			}

			output.push(objectSpec);
		}

		return output;
	}

	/** This method is responsible for getting all objects from a request sample */
	private requestSampleEdgeCase(
		propertyKey: string,
		bindingOptions: IInternalRequestBindOptions,
		currentObjSpec: ObjectDefinitionSpec,
		requestObjSpecs: ObjectDefinitionSpec[],
	) {
		// Get all definitions from this object sample
		const objDefinitions = this.getObjSpecFromSample(
			[{ obj: bindingOptions.specOptions?.sample, parentKey: propertyKey }],
			currentObjSpec.name,
			'Input',
		);

		// Then point the current request definition property to the first definition cretated by the sample
		if (objDefinitions.length > 0) {
			requestObjSpecs.push(...objDefinitions);
			currentObjSpec.properties.push({
				name: propertyKey,
				type: objDefinitions[0].name,
				typeKey: '$ref',
			});
		}
	}

	public mapResponseSamples(requestName: string, request: Record<string, any>): ObjectDefinitionSpec[] {
		const objs: sampleObjContext[] = [];

		// Run through all request's spec options registered on metadata
		for (const responseSpecOptions of specs.responseOptions(request.constructor).sort((s) => +s.status)) {
			if (responseSpecOptions.schema) {
				const statusName = statusNameMap[responseSpecOptions.status] ?? '';

				// The default definition name for the responses will be the request name + status name + Response
				// Example: GetUsersOKResponse, GetUsersNotFoundResponse
				objs.push({
					obj: responseSpecOptions.schema,
					name: `${requestName}${statusName}Response`,
				});
			}
		}

		return this.getObjSpecFromSample(objs, requestName, 'Output');
	}

	private getObjSpecFromSample(objs: sampleObjContext[], requestName: string, suffix: string) {
		// Run through all object entries, if found a new object add it to the stack
		// Based on the value type create the definition object properties

		const output: ObjectDefinitionSpec[] = [];
		while (objs.length > 0) {
			const currentObj = objs.pop()!;

			// The response's objects had their names defined based on status code + 'Response' suffix
			// The inner objects will have their names defined based on the RequestName + parent object property name + suffix
			// Example: GetUsersAddressListOutput, GetUsersZipCodeOutput
			const objectSpec: ObjectDefinitionSpec = {
				name: currentObj.name ?? `${requestName}${capitalize(currentObj.parentKey)}${suffix}`,
				properties: [],
			};

			Object.entries(currentObj.obj).forEach(([key, value]) => {
				if (value !== undefined) {
					let propType = typeof value as string;

					// If the value is an array, it's necessary to check if we have more objects inside it
					if (propType === 'object' && Array.isArray(value)) {
						propType = 'array';

						const firstItem = value[0];
						if (typeof firstItem === 'object') {
							objs.push({
								obj: firstItem,
								parentKey: key,
							});
						}
					}

					const parse = responseSpecParses[propType];
					const input: resParseInput = {
						prop: key,
						value: value,
						requestName: requestName,
						objSuffix: suffix,
					};

					objectSpec.properties.push(parse(input));

					if (propType === 'object' && !Array.isArray(value)) {
						objs.push({
							obj: value,
							parentKey: key,
						});
					}
				}
			});

			output.push(objectSpec);
		}

		return output;
	}

	private getNameFromType(type: any): string {
		return Object.getPrototypeOf(type).constructor.name;
	}

	private isCommandRequest(requestModel: unknown, options: IInternalRequestBindOptions): boolean {
		if (requestModel instanceof PayloadCommandRequest) {
			const commandReq = requestModel as PayloadCommandRequest<unknown, unknown>;
			options.objFactory = commandReq.payloadFactory;
			return true;
		}

		return false;
	}
}

function capitalize(s?: string): string {
	return s ? s.charAt(0).toUpperCase() + s.slice(1) : '';
}
