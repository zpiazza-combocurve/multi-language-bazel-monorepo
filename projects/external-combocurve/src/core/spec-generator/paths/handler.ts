/* eslint-disable 
	complexity,
	@typescript-eslint/explicit-module-boundary-types, 
	@typescript-eslint/no-explicit-any,
	@typescript-eslint/no-non-null-assertion */
import Handlebars from 'handlebars';

import { IInternalRequestBindOptions, RequestBindOptionsKey, RequestBindPropertiesKey } from '@src/core/common';
import { PayloadCommandRequest } from '@src/core/requests/base';
import { specs } from '@src/core/metadata/metadata';

import { Controller, RouteRequest } from '../../controllers/base';
import { EndpointParameterSpec, EndpointPathSpec, EndpointResponseSpec, pathsTemplate } from '../templates';
import { getExample } from '../example.helper';

const statusNameMap: Record<number, string> = {
	200: 'OK',
	201: 'Created',
	202: 'Accepted',
	204: 'NoContent',
	207: 'MultiStatus',
	400: 'BadRequest',
	401: 'Unauthorized',
	403: 'Forbidden',
	404: 'NotFound',
};

export class PathsGenHandler<TController extends Controller> {
	private controller: TController;
	private definitionsBasePath: string;
	private baseRoute: string;
	private idPrefix: string;

	constructor(baseRoute: string, controller: TController, idPrefix?: string, definitionsBasePath?: string) {
		this.baseRoute = baseRoute;
		this.controller = controller;
		this.idPrefix = idPrefix ?? 'v1';
		this.definitionsBasePath = definitionsBasePath ?? '#/definitions/';
	}

	public getPathSpec(): string {
		const output: Record<string, EndpointPathSpec[]> = {};

		for (const routeRegister of this.controller.Routes) {
			const routeEndpoint = this.baseRoute + routeRegister.route;
			if (routeRegister.notGenerateSpecs !== true) {
				const routeSpec = this.getRouteSpec(routeRegister);

				const methodsForRoute = output[routeEndpoint] ?? [];
				methodsForRoute.push(routeSpec);

				output[routeEndpoint] = methodsForRoute;
			}
		}

		const templateInput = {
			routes: Object.entries(output).map(([endpoint, methods]) => ({
				completeRoute: endpoint,
				methods,
			})),
		};

		const template = Handlebars.compile(pathsTemplate);
		const result = template(templateInput);

		return result.substring(result.indexOf('\n') + 1);
	}

	public getRouteSpec(route: RouteRequest): EndpointPathSpec {
		const requestName = getNameFromType(route.requestSample).toLowerCase();
		const routeSpec: EndpointPathSpec = {
			verb: route.method,
			id: `${route.method}-${this.idPrefix}-${requestName}`,
			parameters: [],
			responses: [],
		};

		if (route.method === 'post' || route.method === 'put') {
			routeSpec.consumes = 'application/json';
			routeSpec.produces = 'application/json';
		}

		routeSpec.parameters = [...this.getParameters(route.requestSample)];
		routeSpec.responses = [...this.getResponses(route.requestSample)];

		return routeSpec;
	}

	public *getParameters(requestSample: any): Generator<EndpointParameterSpec> {
		const bindProperties = Reflect.getMetadata(RequestBindPropertiesKey, requestSample) as string[];
		for (const propertyKey of bindProperties ?? []) {
			const options = Reflect.getMetadata(
				RequestBindOptionsKey,
				requestSample,
				propertyKey,
			) as IInternalRequestBindOptions;

			if (!options || options.fromWhere === 'services') {
				continue;
			}

			if (options.fromWhere === 'composition') {
				for (const param of this.getParameters(
					options.objFactory ? options.objFactory() : requestSample[propertyKey],
				)) {
					yield param;
				}

				continue;
			}

			if (options.expects === 'take' || options.expects === 'skip' || options.expects === 'sort') {
				yield this.getSearchParam(requestSample, options);
			}

			const parameter: any = {
				fromWhere: options.fromWhere,
				name: propertyKey,
				required: options.isOptional ? false : true,
				type: options.expects === 'objectID' ? 'string' : options.expects,
				example: getExample(options.expects, propertyKey),
			};

			if (!this.checkPayloadCommandRequestEdgeCase(requestSample, parameter, options)) {
				if (options.expects === 'object') {
					this.completeObjectParameter(propertyKey, requestSample, parameter, options);
				} else if (options.expects === 'array') {
					this.completeArrayParameter(parameter, options);
				}
			}

			yield parameter as EndpointParameterSpec;
		}
	}

	private getSearchParam(requestSample: any, options: IInternalRequestBindOptions): EndpointParameterSpec {
		const searchParam = options.expects as 'take' | 'skip' | 'sort';
		switch (searchParam) {
			case 'skip':
				return {
					fromWhere: 'query',
					name: 'skip',
					description: 'number of items to skip',
					default: requestSample.skip ?? '0',
					type: 'integer',
					format: 'int32',
				};
			case 'take':
				return {
					fromWhere: 'query',
					name: 'take',
					type: 'integer',
					format: 'int32',
					description: 'max records to return',
					default: requestSample.take ?? '25',
					minimum: 1,
					maximum: 200,
				};
			case 'sort':
				return {
					fromWhere: 'query',
					name: 'sort',
					type: 'string',
					default: 'id',
					description:
						'field to sort by, including + or - at the beginning for ascending or descending order, respectively',
					stringEnum: options.requirements?.validValues as string[],
				};
		}
	}

	private checkPayloadCommandRequestEdgeCase(
		requestSample: any,
		parameter: any,
		options: IInternalRequestBindOptions,
	): boolean {
		if (options.isBody && requestSample instanceof PayloadCommandRequest) {
			parameter.isArray = true;
			parameter.itemsTypeKey = '$ref';
			parameter.itemsType = `${this.definitionsBasePath}${getNameFromType(options.objFactory!())}`;
			return true;
		}

		return false;
	}

	private completeObjectParameter(
		propertyKey: string,
		requestSample: any,
		parameter: any,
		options: IInternalRequestBindOptions,
	) {
		delete parameter.type;

		if (options.customParse !== undefined && options.objFactory === undefined) {
			const reqName = getNameFromType(requestSample);
			const propName = capitalize(propertyKey);

			parameter.schema = `${this.definitionsBasePath}${reqName}${propName}Input`;
			return;
		}

		parameter.schema = `${this.definitionsBasePath}${getNameFromType(options.objFactory!())}`;
	}

	private completeArrayParameter(parameter: any, options: IInternalRequestBindOptions) {
		parameter.isArray = true;
		parameter.minItems = options.requirements?.minItems;
		parameter.maxItems = options.requirements?.maxItems;

		delete parameter.example;

		if (options.expects === 'object') {
			const arrayObj = options.objFactory!();

			parameter.itemsTypeKey = '$ref';
			parameter.itemsType = `${this.definitionsBasePath}${getNameFromType(arrayObj)}`;
		} else {
			parameter.itemsTypeKey = 'type';
			parameter.itemsType = options.itemsExpects === 'objectID' ? 'string' : options.itemsExpects;
			parameter.itemsExample = getExample(options.itemsExpects!, '');
		}
	}

	public *getResponses(requestSample: any): Generator<EndpointResponseSpec> {
		for (const responseSpecOptions of specs.responseOptions(requestSample.constructor).sort((s) => +s.status)) {
			const statusName = statusNameMap[responseSpecOptions.status] ?? '';

			const responseSpec: EndpointResponseSpec = {
				status: responseSpecOptions.status,
				description: `${statusName} Response`,
			};

			if (responseSpecOptions.schema) {
				const requestName = getNameFromType(requestSample);
				responseSpec.schema = `${this.definitionsBasePath}${requestName}${statusName}Response`;
			}

			if (responseSpecOptions.headers) {
				responseSpec.headers = Object.entries(responseSpecOptions.headers).map(([key, value]) => ({
					name: key,
					type: typeof value,
				}));
			}

			yield responseSpec;
		}
	}
}

function getNameFromType(type: any): string {
	return Object.getPrototypeOf(type).constructor.name;
}

function capitalize(s?: string): string {
	return s ? s.charAt(0).toUpperCase() + s.slice(1) : '';
}
