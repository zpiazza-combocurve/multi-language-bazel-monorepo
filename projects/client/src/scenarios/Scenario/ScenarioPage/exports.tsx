import { faDownload } from '@fortawesome/pro-regular-svg-icons';
import { omit, pick } from 'lodash-es';
import { useCallback, useState } from 'react';

import { useCallbackRef } from '@/components/hooks';
import { ButtonItem, MenuButton } from '@/components/v2';
import useLDFeatureFlags from '@/feature-flags/useLDFeatureFlags';
import { confirmationAlert, genericErrorAlert, infoAlert } from '@/helpers/alerts';
import { useAlfa } from '@/helpers/alfa';
import { useProgressBar } from '@/helpers/progress';
import { postApi } from '@/helpers/routing';
import { NotificationType } from '@/notifications/notification';
import { useUserNotificationCallback } from '@/notifications/useUserNotificationCallback';
import { generateXlsx } from '@/scenarios/Scenario/ScenarioPage/exports/scenario-table-xlsx';
import { useCurrentScenario } from '@/scenarios/api';
import { ASSUMPTION_KEYS, ASSUMPTION_NAMES_BY_KEY, NON_SCENARIO_ASSUMPTION_KEYS } from '@/scenarios/shared';

import {
	CCImportDialog,
	ExportProbabilisticDialog,
	ExportToAriesDialog,
	ExportToCsvDialog,
	ExportToPhdwinDialog,
	useCCExport,
} from './exports/index';

const INITIAL_PROGRESS = 1;

export function useExportDialogProps() {
	const [visible, setVisible] = useState(false);
	const close = () => setVisible(false);
	const open = () => setVisible(true);
	return { close, open, visible };
}

export function useXLSXDownload({ headers }) {
	const { wellHeaders } = useAlfa();
	const { scenario } = useCurrentScenario();
	const [progress, updateProgress] = useState<null | number>(null);
	useProgressBar(progress);
	const download = async () => {
		infoAlert('Export started. Please, do not close your browser.');
		updateProgress(INITIAL_PROGRESS);

		generateXlsx({
			assumptionKeysList: ASSUMPTION_KEYS.filter((e) => !NON_SCENARIO_ASSUMPTION_KEYS.includes(e)),
			scenario, // TODO receive from props
			wellHeaders,
			tableHeaders: headers,
			assumptionNamesByKey: omit(ASSUMPTION_NAMES_BY_KEY, NON_SCENARIO_ASSUMPTION_KEYS),
			updateProgress,
		});
	};
	return { download };
}

export function useTableWithLookupDownload({ scenarioId, headers, assignments }) {
	const {
		wellHeaders,
		user: { _id: userId },
	} = useAlfa();
	const download = async () => {
		const tableWellHeaders = {
			...pick(wellHeaders, headers),
			well_name: 'Well',
		};
		const finalAssumptionKeys = omit(ASSUMPTION_NAMES_BY_KEY, NON_SCENARIO_ASSUMPTION_KEYS);
		const body = {
			user: userId,
			selectedAssignmentIds: assignments,
			assumptionKeys: finalAssumptionKeys,
			wellHeaders: tableWellHeaders,
		};

		try {
			await postApi(`/scenarios/${scenarioId}/scenarioTableWithLookup`, body);
		} catch (err) {
			genericErrorAlert(err);
		}
	};
	return { download };
}

export function useExportOptions({
	scenarioId,
	scenarioName,
	selectedHeaders,
	selectedAssignmentIds,
	sortedFilteredAssignmentIds,
}) {
	const { Pusher } = useAlfa();
	const sharedExportDialogProps = {
		Pusher,
		scenarioId,
		scenarioName,
		headers: selectedHeaders,
		selectedAssignmentIds,
		// eslint-disable-next-line new-cap -- TODO eslint fix later
		timeZone: Intl.DateTimeFormat().resolvedOptions().timeZone,
	};

	const sharedDownloadProps = {
		scenarioId,
		headers: selectedHeaders,
		assignments: sortedFilteredAssignmentIds,
	};

	const { open: openCsvExportDialog, ...csvExportProps } = useExportDialogProps();
	const { open: openAriesExportDialog, ...ariesExportProps } = useExportDialogProps();
	const { open: openPhdwinExportDialog, ...phdwinExportProps } = useExportDialogProps();
	const { open: openExportProbabilisticDialog, ...exportProbabilisticProps } = useExportDialogProps();
	const { download: downloadXlsx } = useXLSXDownload(sharedDownloadProps);
	const { download: exportWithLookup } = useTableWithLookupDownload(sharedDownloadProps);

	const { isProbabilisticCapexEnabled } = useLDFeatureFlags();

	const menuItems = [
		{
			primaryText: 'Mass Export Assumptions',
			onClick: openCsvExportDialog,
			disabled: !selectedAssignmentIds.length,
		},
		{
			primaryText: 'Export Model Names To Excel',
			onClick: downloadXlsx,
		},
		{
			primaryText: 'Export Rendered Lookup Tables to CSV',
			onClick: exportWithLookup,
		},
		{
			primaryText: 'Export Scenario To ARIES',
			onClick: openAriesExportDialog,
		},
		{
			primaryText: 'Export Assumptions To PHDWIN',
			onClick: openPhdwinExportDialog,
		},
	];

	if (isProbabilisticCapexEnabled) {
		menuItems.push({
			primaryText: 'Generate Probabilistic Inputs',
			onClick: openExportProbabilisticDialog,
		});
	}

	const exportMenu = (
		<MenuButton label='EXPORTS' startIcon={faDownload}>
			{menuItems.map(({ primaryText, onClick, disabled }) => (
				<ButtonItem label={primaryText} onClick={onClick} disabled={disabled} key={primaryText} />
			))}
		</MenuButton>
	);

	const exportToCsvDialog = (
		<ExportToCsvDialog {...sharedExportDialogProps} tableHeaders={selectedHeaders} {...csvExportProps} />
	);
	const exportToAriesDialog = <ExportToAriesDialog {...sharedExportDialogProps} {...ariesExportProps} />;
	const exportToPhdwinDialog = <ExportToPhdwinDialog {...sharedExportDialogProps} {...phdwinExportProps} />;
	const exportProbabilisticDialog = (
		<ExportProbabilisticDialog {...sharedExportDialogProps} {...exportProbabilisticProps} />
	);

	return { exportMenu, exportToCsvDialog, exportToAriesDialog, exportToPhdwinDialog, exportProbabilisticDialog };
}

export function useColumnExports({ selectedAssignmentIds, scenarioId, scenarioName, selectedHeaders, refetch }) {
	const { scenario, update: updateScenario } = useCurrentScenario();
	// eslint-disable-next-line @typescript-eslint/no-explicit-any -- TODO eslint fix later
	const [ccImportDialog, setCCImportDialog] = useState<any>({ visible: false });
	const [exportDialog, promptExportDialog] = useCCExport();
	// eslint-disable-next-line @typescript-eslint/no-explicit-any -- TODO eslint fix later
	const [importCSVNotificationCallback, setImportCSVNotificationCallback] = useState<any>(null);

	useUserNotificationCallback(NotificationType.CC_CC_IMPORT, importCSVNotificationCallback);

	const exportCSV = useCallbackRef(async ({ assumptionKey }) => {
		const assumptionName = ASSUMPTION_NAMES_BY_KEY[assumptionKey];
		promptExportDialog({
			scenarioId,
			scenarioName,
			selectedAssignmentIds,
			tableHeaders: selectedHeaders,
			assumptionKey,
			assumptionName,
		});
	});
	const importCSV = useCallback(async ({ assumptionKey }) => {
		const assumptionName = ASSUMPTION_NAMES_BY_KEY[assumptionKey];
		setCCImportDialog({ visible: true, assumptionKey, assumptionName });
	}, []);

	const handleChangeQualifier = useCallback(
		async (column, key) => {
			const newScenario = await postApi(`/scenarios/${scenarioId}/activateQualifier`, { column, key });
			// eslint-disable-next-line @typescript-eslint/no-explicit-any -- TODO eslint fix later
			updateScenario({ ...scenario, columns: newScenario.columns } as any);
			confirmationAlert(`Qualifier applied successfully.`);
			refetch();
		},
		[refetch, scenario, scenarioId, updateScenario]
	);
	const importDialog = (
		<CCImportDialog
			{...{
				ccImportDialog,
				scenarioId,
				hideDialog: () => setCCImportDialog({ visible: false }),
				buildCurrent: refetch,
				handleChangeQualifier,
				refetch,
				setCallback: setImportCSVNotificationCallback,
			}}
		/>
	);
	const dialogs = (
		<>
			{exportDialog}
			{importDialog}
		</>
	);
	return { importCSV, exportCSV, dialogs };
}
