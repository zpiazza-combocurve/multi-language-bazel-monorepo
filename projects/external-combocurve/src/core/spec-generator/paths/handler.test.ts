import { TestController } from '@src/core/test.exports';

import { PathsGenHandler } from './handler';

describe('spec handler', () => {
	it('mapRequestObjects', () => {
		const handler = new PathsGenHandler<TestController>('test_route', new TestController());

		const paths = handler.getPathSpec();

		expect(paths).toMatchSnapshot();
	});
});
