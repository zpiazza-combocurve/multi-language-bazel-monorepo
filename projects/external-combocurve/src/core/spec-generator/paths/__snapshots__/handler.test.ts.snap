// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spec handler mapRequestObjects 1`] = `
"test_route/:
  delete:
    operationId: delete-v1-testmodel
    parameters:
      - name: name
        in: body
        required: true
        type: string
        example: some_string
      - name: age
        in: query
        type: number
        example: 42
      - name: isCool
        in: body
        required: true
        type: boolean
        example: true
      - name: id
        in: params
        required: true
        type: string
        example: 5e276e31876cd70012ddf3f6
      - name: compositionTest
        in: query
        type: string
        example: some_string
      - name: inner
        in: body
        required: true
        schema:
          $ref: '#/definitions/InnerModel'
      - name: address
        in: body
        type: string
        example: some_string
      - name: custom
        in: body
        required: true
        schema:
          $ref: '#/definitions/TestModelCustomInput'
    responses:
      '201':
        description: Created Response
        schema:
          $ref: #/definitions/TestModelCreatedResponse
      '404':
        description: NotFound Response
        schema:
          $ref: #/definitions/TestModelNotFoundResponse
      '200':
        description: OK Response
        schema:
          $ref: #/definitions/TestModelOKResponse
      '401':
        description: Unauthorized Response
      '400':
        description: BadRequest Response
"
`;
